module/builder/collection/builder.go

```diff
@@ -72,35 +73,41 @@ func (b *Builder) BuildOn(parentID flow.Identifier, setter func(*flow.Header) er
 
 	// first we construct a proposal in-memory, ensuring it is a valid extension
 	// of chain state -- this can be done in a read-only transaction
-	err := b.db.View(func(tx *badger.Txn) error {
+	err := b.db.View(func(btx *badger.Txn) error {
```

Comment:
This `View` transaction now contains lots of db queries and other operation. It will increase the chance of race condition and failure. 

I would suggest to break this view transaction into multiple ones. And move some operations out of this view transaction, such as removing txs from transactions mempool

Reply:
👍  Good point, there is a lot here that doesn't need to be in the transaction: updated in https://github.com/onflow/flow-go/pull/2385/commits/c15e68b2ff74b323d21c8836ea1043d15b9603f4